light-client-detector/src/detect.rs

```diff
@@ -0,0 +1,228 @@
+        // No divergence found
+        Ok(()) => Ok(None),
+
+        // We have conflicting headers. This could possibly imply an attack on the light client.
+        // First we need to verify the witness's header using the same skipping verification and then we
+        // need to find the point that the headers diverge and examine this for any evidence of an attack.
+        //
+        // We combine these actions together, verifying the witnesses headers and outputting the trace
+        // which captures the bifurcation point and if successful provides the information to create valid evidence.
+        Err(CompareError::ConflictingHeaders(challenging_block)) => {
+            warn!(
+                witness = %witness.peer_id(),
+                height  = %challenging_block.height(),
+                "Found conflicting headers between primary and witness"
+            );
+
+            // Gather the evidence to report from the conflicting headers
+            let evidence = gather_evidence_from_conflicting_headers(
+                None,
+                witness,
+                &primary_trace,
+                &challenging_block,
+            )
+            .await?;
+
+            Ok(Some(Divergence {
+                evidence: evidence.against_primary,
+                challenging_block: *challenging_block,
+            }))
+        },
+
+        Err(CompareError::BadWitness) => {
+            // These are all melevolent errors and should result in removing the witness
+            debug!(witness = %witness.peer_id(), "witness returned an error during header comparison, removing...");
+
+            Err(Error::bad_witness())
+        },
+
+        Err(CompareError::Other(e)) => {
+            // Benign errors which can be ignored
+            debug!(witness = %witness.peer_id(), "error in light block request to witness: {e}");
+
+            Err(Error::light_client(e))
+        },
+    }
+}
+
+#[derive(Debug)]
+pub enum CompareError {
+    BadWitness,
+    Other(tendermint_light_client::errors::Error),
+    ConflictingHeaders(Box<LightBlock>),
+}
+
+/// compareNewHeaderWithWitness takes the verified header from the primary and compares it with a
+/// header from a specified witness. The function can return one of three errors:
+///
+/// 1: errConflictingHeaders -> there may have been an attack on this light client
+/// 2: errBadWitness -> the witness has either not responded, doesn't have the header or has given us an invalid one
+///
+/// Note: In the case of an invalid header we remove the witness
+///
+/// 3: nil -> the hashes of the two headers match
+pub fn compare_new_header_with_witness(
+    new_header: &SignedHeader,
+    witness: &mut Provider,
+    max_clock_drift: Duration,
+    max_block_lag: Duration,
+) -> Result<(), CompareError> {
+    let light_block = check_against_witness(new_header, witness, max_clock_drift, max_block_lag)?;
+
+    if light_block.signed_header.header.hash() != new_header.header.hash() {
```

Comment:
I am doing it here so that I don't have to duplicate the check in the three places where we return `Ok(light_block)` in `check_against_witness`. I'll see if I can cleanly move it there.